src/App.jsx

```diff
@@ -1,65 +1,132 @@
 
-const LOCALSTORAGE_KEY = 'savedBlogTitles'
+const initialTitles = [
+  'Lesson 1.1 Project setup',
+  'Lesson 1.2 React DOM and components',
+  'Lesson 1.3 Form input, props, state and callbacks',
+  'Lesson 1.4 Lifting state, controlled components, props handling',
+  'Lesson 1.5 Side effects, custom hooks, fragments',
+  'Lesson 1.6 Reusable Components, Imperative React',
+  'Lesson 1.7 Asynchronous Data, Conditional Rendering',
+];
+
+const LOCALSTORAGE_KEY = 'savedBlogTitles';
+
+const filterBlogTitles = (blogTitles, searchText) =>
+  blogTitles.filter(blogTitle =>
+    blogTitle.title.toLowerCase().includes(searchText.toLowerCase())
+  );
+
+const stateManagementFunction = (previousState, action) => {
+  switch (action.type) {
+    case 'FINISH_LOADING_BLOG_TITLES':
+      return {
+        ...previousState,
+        isLoading: false,
+        blogTitles: action.payload.blogTitles,
+      };
+    case 'SET_SEARCH_TEXT':
+      return {
+        ...previousState,
+        searchText: action.payload.searchText,
+      };
+    case 'START_ADDING_BLOG_TITLE':
+      return {
+        ...previousState,
+        isAdding: true,
+      };
+    case 'FINISH_ADDING_BLOG_TITLE':
+      return {
+        ...previousState,
+        isAdding: false,
+        blogTitles: [
+          ...previousState.blogTitles,
+          titleToBlog(action.payload.newTitle),
+        ],
+        focus: 'addBlog',
+      };
+    case 'RESET_FOCUS':
+      return {
+        ...previousState,
+        focus: null,
+      };
+  }
+
+  // return a new state, that is the entire state
+};
+
+// initial state should have the same properties as all new states returned
+const initialState = {
+  blogTitles: [],
+  isLoading: true,
+  isAdding: false,
+  searchText: '',
+  focus: 'search',
+};
 
 function App() {
-    const initialTitles = ["Lesson 1.1 Project setup",
-        "Lesson 1.2 React DOM and components",
-        "Lesson 1.3 Form input, props, state and callbacks",
-        "Lesson 1.4 Lifting state, controlled components, props handling",
-        "Lesson 1.5 Side effects, custom hooks, fragments",
-        "Lesson 1.6 Reusable Components, Imperative React",
-        "Lesson 1.7 Asynchronous Data, Conditional Rendering"
-    ]
-    const initialBlogs = initialTitles.map(title => titleToBlog(title))
-    const [blogTitles, setBlogTitles] = useState([])
-    const [isLoading, setIsLoading] = useState(true)
+  const initialBlogs = initialTitles.map(title => titleToBlog(title));
+  const [state, runAction] = useReducer(stateManagementFunction, initialState);
 
-    useEffect(() => {
-        new Promise(
-            (resolve, reject) => {
-                setTimeout(() => {
-                    resolve(getItem(LOCALSTORAGE_KEY) || initialBlogs)
-                }, 2000)
-            }
-        ).then((blogTitlesFromLocalStorage) => {
-            setBlogTitles(blogTitlesFromLocalStorage)
-            setIsLoading(false)
```

Comment:
The FINISH_LOADING_BLOG_TITLES action helps us organize our state into declarative operations. It's much easier to identify what the state should be in the reducer by looking at the returned object rather than reading several operations of setting independent state.